E2e protocol ricardo 20231025 login sync

corrideat opened the pull request.

corrideat committed:
Login sync fixes

corrideat committed:
Fix various issues related to login and re-joining

corrideat committed:
Make saveOurLoginState async

corrideat committed:
Various improvments and bugfixes

Silver-IT commented:
Having checked this PR mainly focusing on the user flow, I could see several issues.

**Issue 1**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group
  - [Tab2] user2 joins the group, and closes the tab
  - [Tab1] user1 kicks the u2 from the group
  - [Tab2] Opens 'http://localhost:8000'. Got 5 errors in the console
  - [Tab1] Got error in console too.
- Screenshot
  - Error in Tab1
![image](https://github.com/okTurtles/group-income/assets/57976479/4dbcc99d-efb6-4a1e-a747-5b1c5eb61b1c)
  - 5 Errors in Tab2
![image](https://github.com/okTurtles/group-income/assets/57976479/76929647-25ea-47d7-8ab8-2e3a318494c7)
![image](https://github.com/okTurtles/group-income/assets/57976479/a10437e0-0843-4cda-848f-6c473ee1701e)
![image](https://github.com/okTurtles/group-income/assets/57976479/ed8535af-5052-45ff-b4ee-4deb40cd5da6)
![image](https://github.com/okTurtles/group-income/assets/57976479/be404689-d29b-434d-bcfa-62f9fe325909)
- What should happen
  - No errors should be happening. And the important thing is that there MUST be no errors in Tab1.
  - `gi.actions/group/join` shouldn't be called. But it is called twice.
- DeadLock Output
  - [Tab1] {}
  - [Tab2] {}

Silver-IT commented:
**Issue 2**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group
  - [Tab2] user2 joins the group, and closes the tab
  - [Tab1] user1 kicks the u2 from the group
  - [Tab3] user2 logs in, and got 5 errors.
- Screenshot
  - 5 Errors in Tab2
![image](https://github.com/okTurtles/group-income/assets/57976479/2f8cc1bc-d700-473a-adf6-05b7ba7d43ce)
![image](https://github.com/okTurtles/group-income/assets/57976479/fb2895bb-ced1-4e22-a5ad-b17543776b0e)
![image](https://github.com/okTurtles/group-income/assets/57976479/50a51b0f-612a-4107-86a5-44ca0a5fccd2)
![image](https://github.com/okTurtles/group-income/assets/57976479/695ce6d7-d190-4b62-b82e-67b08d80600b)
![image](https://github.com/okTurtles/group-income/assets/57976479/3f83a465-a739-4529-8de6-fce37dba5e61)
- What should happen
  - No errors should be happening.
  - `gi.actions/group/join` should be called ONLY once. But it is called twice.
- DeadLock Output
  - [Tab1] {}
  - [Tab2] {}

Silver-IT commented:
**Issue 3**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group.
  - [Tab2] user2 joins the group.
  - [Tab1] user1 kicks the u2 from the group. Got 1 error in console.
  - [Tab2] user2 joins the group again with the initial invite link.
  - [Tab2] Wrong work flow to join. 6 Errors in console. Seems there are many redundant function calls.
  - [Tab1] Many logs in console than expected. Seems there are too many function calls. And duplicated notifications.
- Screenshot
  - Error in Tab1 (Step 3)
![image](https://github.com/okTurtles/group-income/assets/57976479/91126036-a8bc-43ce-b896-8f07de658a35)
  - Errors in Tab2 (Step 5)
![image](https://github.com/okTurtles/group-income/assets/57976479/ca9f5e1c-80a3-4b03-9a05-6dd41d9f5171)
![image](https://github.com/okTurtles/group-income/assets/57976479/a9e4667c-68f2-458f-be81-5804d8d545cc)
  - Duplicated notifications in Tab1 (Step 6)
![image](https://github.com/okTurtles/group-income/assets/57976479/dac944db-9c91-454c-a811-b1c2db94cfc1)
- What should happen
  - No errors should be happening.
  - No duplicated notifications in Tab1
  - `gi.actions/group/join` should be called ONLY once. But it is called 4 times.
- DeadLock Output
  - [Tab1]
![image](https://github.com/okTurtles/group-income/assets/57976479/4afe5475-6e9e-49c0-908e-b3f957a05ec1)
  - [Tab2] {}

Silver-IT commented:
**Issue 4**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group.
  - [Tab2] user2 joins the group.
  - [Tab3] user3 joins the group, and closes the tab.
  - [Tab2] user2 creates a proposal to remove user3 from the group.
  - [Tab1] user1 approves the proposal, and kicks user3 from the group. Got an error in the notifications.
  - [Tab4] user3 tries to log in, and 8 errors BANG.
- Screenshot
  - Error in Tab1 (Step 5)
(user1 = alexjin, user2 = greg, user3 = andrea)
![image](https://github.com/okTurtles/group-income/assets/57976479/6f5abbca-4689-40ce-9e19-a657c91222fc)
  - Errors in Tab4 (Step 6)
![image](https://github.com/okTurtles/group-income/assets/57976479/368b7975-79b3-4c1b-a49d-d85133cb7afa)
![image](https://github.com/okTurtles/group-income/assets/57976479/41321b94-9385-4232-973f-54ed78a32994)
![image](https://github.com/okTurtles/group-income/assets/57976479/ca99b4ef-8548-40e7-8623-aec03a645256)
![image](https://github.com/okTurtles/group-income/assets/57976479/b99cab13-a1d4-4d05-9143-46c262548e2b)
![image](https://github.com/okTurtles/group-income/assets/57976479/65eded34-7cf0-4fe5-9a64-c8f5aa0e3b68)
![image](https://github.com/okTurtles/group-income/assets/57976479/908aabab-7d60-4fa2-b813-610e404cd153)
![image](https://github.com/okTurtles/group-income/assets/57976479/0171a841-78a1-4860-bb24-99b33343d0f7)
![image](https://github.com/okTurtles/group-income/assets/57976479/543159be-520b-4c3d-91fd-79da99e01c5e)
- What should happen
  - No errors should be happening.
  - `gi.actions/group/join` should be called ONLY once. But it is called twice in Tab4 (Step 6).
- DeadLock Output
  - [Tab1]
![image](https://github.com/okTurtles/group-income/assets/57976479/e9fad5d5-5240-41e6-9ebe-946fc28d7ef0)
  - [Tab2] {}
  - [Tab4] {}

Silver-IT commented:
**Issue 5**
- Steps to reproduce
  - [Tab1] user1 signs up, and creates a group.
  - [Tab2] user2 joins the group.
  - [Tab3] user1 logs in.
  - [Tab4] user2 logs in.
  - [Tab1] user1 kicks user2 from the group.
  - [Tab1], [Tab2], [Tab4] Errors BANG.
- Screenshot
  - Error2 in Tab1 (Step 6)
(user1 = alexjin, user2 = greg)
![image](https://github.com/okTurtles/group-income/assets/57976479/fd88c403-151f-42de-8c2d-a882834ea418)
  - Errors in Tab2 (Step 6)
![image](https://github.com/okTurtles/group-income/assets/57976479/24d9fcfd-20ba-4a99-8b91-7d336c4ac686)
  - Errors in Tab4 (Step 6)  
![image](https://github.com/okTurtles/group-income/assets/57976479/370a335b-b4f9-4a46-a2b8-d0328599b2b7)
- What should happen
  - No errors should be happening.
- DeadLock Output
  - [Tab1] {}
  - [Tab2] {}
  - [Tab3] {}
  - [Tab4] {}

corrideat committed:
Bugfixes. Revert changes to side-effects

corrideat commented:
Thanks again, @Silver-IT!

As discussed:

1. I've made side-effects mostly work as they did earlier
2. I've added a check on `syncContractAndWatchKeys` to avoid syncing contracts that have been subscribed to

In addition, I've implemented fixes for all of the issues reported here, except issue 3. I'll continue working on this until issue 3 is resolved.

Regarding issue 1, there is the double sync, that I've tracked down to the login functionality. I'll work on addressing this as well (but it seems to be the case that implementing the `join|leaveGroup` actions would help solving it too).

@taoeffect: You'll probably be pleased to see that I've mostly removed `sync` from `recreateEvent`, something you weren't entirely happy with. There are some vestiges of it that remain, because _sometimes_ the `pubsub` functions don't work as expected and some events are never received. Once this is addressed in `pubsub`, the `syncContract` added to `publishEvent` can be removed.

corrideat committed:
bump

corrideat committed:
Address re-joining of chatrooms

corrideat commented:
@Silver-IT I've made another commit addressing issue 3. The solution so far is using `sbp('okTurtles.data/set', 'JOINING_CHATROOM-' + params.contractID, ?)` to avoid removing the chatroom contract when joining.

Besides that, two issues remain.

One is at _leaving_; specifically, when someone else leaves a chatroom we have left:

To write to a contract, we need to be subscribed to it. This is done in `encryptedAction`, but for contracts that call `/remove` in a side-effect, this is essentially pointless. This is an error.

The second issue is at _joining_. You'll see that the `/join` action calls `/sync`. This is 'unnecessary', as `encryptedAction` already does this. However, similarly to what happened above, the first `/sync` will be pointless because of the side-effect removing the contract, requiring a second call to sync that sets the `JOINING_CHATROOM-${cID}` key.

I think that we need to either adapt `encryptedAction` to take pre- and post-hooks (we could call them, e.g., prologue and epilogue) to handle this, or somehow re-think how `/remove` gets called (I think that one thing that would work is doing reference counting, but there may be other options too). These are not mutually exclusive, so we could very well do both.

Related to this, there is a third issue, which is a variant of issue 3:

* rejoin fails 
  - [tab1] u1: create group
  - [tab2] u2: join using invite link
  - [tab2] u2: close tab / navigate away
  - [tab1] u1: remove u2
  - [tab2] u2: open tab with invite link

To recap, apart from the issues mentioned, which are related to removing contracts, there is the double-sync issue I mentioned earlier.

Although double-syncs are a concern, I'll focus now on trying to address the actual issues that happen from missing contract states, which are errors.

Do you have a suggestion on how we could call `/remove` in an elegant way?

Silver-IT commented on frontend/model/contracts/group.js:
```diff
@@ -938,101 +985,17 @@ sbp('chelonia/defineContract', {
+        // re-joining works
+        sbp('chelonia/queueInvocation', contractID, () => sbp('gi.contracts/group/leaveGroup', { data, meta, contractID }, { state, getters })).catch(e => {
+          console.error(`[gi.contracts/group/removeMember/sideEffect] Error ${e.name} during queueInvocation for ${contractID}`, e)
         })
```
We don't need to call `gi.contracts/group/leaveGroup` when the member has joined the group after this action.
So it would be necessary to check the group contract profile state in this invocation.
Please reference [my PR](https://github.com/okTurtles/group-income/pull/1728) for this.

```
sideEffect ({ data, meta, contractID }, { state, getters }) {
  const rootState = sbp('state/vuex/state')
  const rootGetters = sbp('state/vuex/getters')
  const contracts = rootState.contracts || {}
  const { username } = rootState.loggedIn

  if (data.member === username) {
    // If this member is re-joining the group, ignore the rest
    // so the member doesn't remove themself again.
    sbp('chelonia/queueInvocation', contractID, async () => {
      if (rootState[contractID].profiles?.[username]?.status === PROFILE_STATUS.REMOVED) {
        // NOTE: should remove archived data from IndexedStorage
        //       regarding the current group (proposals, payments)
        await sbp('gi.contracts/group/removeArchivedProposals', contractID)
        await sbp('gi.contracts/group/removeArchivedPayments', contractID)
```

Silver-IT commented on frontend/model/contracts/group.js:
```diff
@@ -261,38 +261,79 @@ function updateGroupStreaks ({ state, getters }) {
-    }),
+    ...extraParams,
     hooks: {
       preSendCheck: (_, state) => {
         // Avoid sending a duplicate action if the person has already
         // left or been removed from the chatroom
         return !!state?.users?.[member]
       }
```
Inside `group` contract, it doesn't make sense to decide the next operation depending on the `chatroom` contract state.

Silver-IT reviewed (changes requested):
Nice work, @corrideat. I did a very partial review, and left a just two comments.
I can see many updates in contract parts (model/contracts, controller/actions) but I don't think most of them are needed. I have almost finished my PR #1728 and I want you to have even a brief look at it. I fixed most of the login issues with the smaller changes than this.

Also will check the 5 Issues (described above) again, and left comments here.

Silver-IT commented:
**Issue 1** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784531132) This issue is not fixed yet.
![image](https://github.com/okTurtles/group-income/assets/57976479/ef874400-6ee7-4dd3-8722-3c2bc304703e)
This screenshot shows the console logs filtered by `gi.actions/group` on Tab2 in Step 5 .

**Problem: three errors and 2 redundant `gi.actions/group/join` function calls**

Silver-IT commented:
**Issue 2** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784557201) This issue is not fixed yet.
![image](https://github.com/okTurtles/group-income/assets/57976479/8d0ff33f-6935-4d00-a839-9b6e93ab5860)
This screenshot shows the console logs filtered by `gi.actions/group/join` on Tab3 in Step 4.

**Problem: 1 redundant `gi.actions/group/join` function call.**

Silver-IT commented:
Issue 3 (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784603749) seems not to be fixed yet.
![image](https://github.com/okTurtles/group-income/assets/57976479/2054c464-feb1-4415-9fb4-1701ad5d41f7)
![image](https://github.com/okTurtles/group-income/assets/57976479/a11f4c78-1d44-4ea8-af1d-b8d540d5a252)
![image](https://github.com/okTurtles/group-income/assets/57976479/2da07a41-6674-4c6e-912a-5fc90a34e65f)

**Problem: two errors in console**

taoeffect commented on shared/domains/chelonia/GIMessage.js:
```diff
@@ -358,25 +358,24 @@ export class GIMessage {
       }
+    } catch {
+      // Signature or encryption error
+      return undefined
```
No `console.warn` is needed here?

corrideat replied:
This is doing the same as was it was doing earlier, more or less. There could be a `console.warn`, but that is already done elsewhere and it feels repetitive to warn multiple times.

taoeffect replied:
Then a comment about that should be here...

taoeffect commented on shared/domains/chelonia/GIMessage.js:
```diff
@@ -407,7 +406,7 @@ export class GIMessage {
   }
 
-  isFirstMessage (): boolean { return !this.head().previousHEAD }
+  isFirstMessage (): boolean { return !this.head().contractID }
```
Why was this changed?

corrideat replied:
Because I removed a bunch of the repetitive code in `chelonia.js` that was looking up the previous HEAD, which is now done by `publishEvent`. Since the previous HEAD could now momentarily be null, the check as it was would fail. Note that either condition is correct (if the previous HEAD were set).

taoeffect commented on frontend/views/pages/PendingApproval.vue:
```diff
@@ -36,7 +37,7 @@ export default ({
     ourGroupProfile () {
       if (!this.ephemeral.groupIdWhenMounted) return
-      return this.$store.state[this.ephemeral.groupIdWhenMounted]?.profiles?.[this.ourUsername]
+      return this.$store.state[this.ephemeral.groupIdWhenMounted]?.profiles?.[this.ourUsername]?.status === PROFILE_STATUS.ACTIVE
```
This change changes the meaning of this computed property.

The computed property is named `ourGroupProfile`, which implies it returns our group profile. But now this no longer happens, instead it returns a boolean.

So either the condition needs to be removed or the name of the computed property needs to be changed.

corrideat replied:
Agreed.

taoeffect reviewed (changes requested):
Preliminary review

taoeffect commented on frontend/common/errors.js:
```diff
@@ -23,3 +23,14 @@ export class GIErrorUIRuntimeError extends Error {
+      Error.captureStackTrace(this, this.constructor)
+    }
+  }
+}
```
I think this is a good time to copy over that utility function from Chelonia to define these using 1-liners...

corrideat replied:
Agreed.

taoeffect commented on shared/domains/chelonia/internals.js:
```diff
@@ -292,14 +318,12 @@ export default (sbp('sbp/selectors/register', {
+        // a call to sync and the subscription time. This is a temporary measure
+        // to handle this until [pubsub] is updated.
+        if (entry.height() === lastAttemptedHeight) {
+          await sbp('okTurtles.eventQueue/queueEvent', entry.contractID(), ['chelonia/private/in/syncContract', entry.contractID()])
```
So I agree that we should address this in this PR... one way to go about it is to make sure the pubsub is connected before `'chelonia/contract/sync'` does the sync.

We might also want to verify that we are in fact connected to the pubsub server before allowing messages to be sent.

taoeffect commented on frontend/model/contracts/group.js:
```diff
@@ -261,38 +261,79 @@ function updateGroupStreaks ({ state, getters }) {
+            memberKey, member, profile, active: (profile: any)?.status === PROFILE_STATUS.ACTIVE
+          })
+          if (memberKey === member && (profile: any)?.status === PROFILE_STATUS.ACTIVE) {
+            return [memberKey, { status: PROFILE_STATUS.REMOVED }]
```
This line seems to be removing many profile fields... that seems wrong.

corrideat replied:
There isn't much in the profile _and_ this is for removed members. Would you prefer `{ ...profile, status: PROFILE_STATUS.REMOVED }` instead?

taoeffect replied:
Yes, I think that would be much more robust.

taoeffect commented on frontend/model/contracts/group.js:
```diff
@@ -261,38 +261,79 @@ function updateGroupStreaks ({ state, getters }) {
+
+  const extraParams = {}
+
+  // When a group is being left, we want to also leave chatrooms,
+  // including private chatrooms. Since the user issuing the action
+  // may not be a member of the chatroom, we use the group's CSK
+  // unconditionally in this situation, which should be a key in the
+  // chatroom (either the CSK or the groupKey)
+  if (leavingGroup) {
+    const signingKeyId = sbp('chelonia/contract/currentKeyIdByName', state, 'csk', true)
+
+    // If we don't have a CSK, it is because we've already been removed.
+    // Proceeding would cause an error
```
These are good comments 👍

taoeffect commented on frontend/model/contracts/group.js:
```diff
@@ -261,38 +261,79 @@ function updateGroupStreaks ({ state, getters }) {
+    }
+
+    extraParams.signingKeyId = signingKeyId
+    extraParams.innerSigningContractID = null
```
Does `innerSigningContractID = null` need to be specified? If yes, could you add a comment explaining why?

corrideat replied:
Yes, it needs to be specified or else it'll be set to the identity CSK.

taoeffect commented on frontend/model/contracts/group.js:
```diff
@@ -261,38 +261,79 @@ function updateGroupStreaks ({ state, getters }) {
     }
+  }).catch((e) => {
+    console.log('jjj action', { e })
+    if (leavingGroup && e?.name === 'GIErrorUIRuntimeError' && e?.cause?.name === 'GIErrorMissingSigningKeyError') {
```
~~This `.cause` thing doesn't seem to be a thing? `GIErrorMissingSigningKeyError` sets `.name` to `GIErrorMissingSigningKeyError`.~~

EDIT: nevermind, it's set in `actions/utils.js`

taoeffect commented on shared/domains/chelonia/utils.js:
```diff
@@ -498,12 +485,13 @@ export const recreateEvent = async (entry: GIMessage, rootState: Object): Promis
 
+  const newOp = opT === GIMessage.OP_ATOMIC && (newRawOpV: any).length === 1
+    ? (newRawOpV: any)[0]
+    : [opT, newRawOpV]
```
What's going on here? Are you sure this is right? Is `? (newRawOpV: any)[0]` correct?

Why is this being done? Is it rewriting what the user is sending? We should probably keep the developer intent and not modify messages the developer is sending.

corrideat replied:
Reverted

taoeffect commented on shared/domains/chelonia/internals.js:
```diff
@@ -1339,64 +1362,66 @@ const handleEvent = {
         }
-        return acc
       }
+      return Promise.resolve(sbp(`${manifestHash}/${action}/sideEffect`, mutation))
```
Why is this wrapping the result in `Promise.resolve`? What if an error occurred? What's the point of later checking `.then` and `Promise.allSettled` if they're guaranteed to settled because of the `Promise.resolve`?

Or is that not how it works? If so please add a comment explaining because this is confusing...

corrideat replied:
Agreed

corrideat committed:
Address PR review feedback

corrideat committed:
Deferred remove

corrideat committed:
Fixes to "chelonia/private/deleteRevokedKeys"

corrideat committed:
Re-add sync to publishEvent

corrideat committed:
Bugfix: reducer at deleteRevokedKeys

corrideat committed:
Test state.contracts[cID] instead of state[cID]

Silver-IT commented:
- Still existing issues:
  - **Issue 1**(https://github.com/okTurtles/group-income/pull/1780#issuecomment-1797865986)
  - **Issue 2** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1797876669)
  - **Issue 3** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1797889651)
  - **Issue 4** (https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784699012)
  Still two function `gi.actions/group/join` calls

corrideat committed:
Bugfixes:

  * Pass params from sync to syncContract
  * Use fresh state in signData and encryptData
  * Check membership status in Join.vue

corrideat committed:
Move joining #General and adding group key into side-effect

Silver-IT commented:
@corrideat, @taoeffect. I have just checked the test scenario.
### Still existing issues in terms of workflow
  - [**Issue 1**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784531132)
  Problem:
    - user2 has successfully logged in.
    - One error occurred.
![image](https://github.com/okTurtles/group-income/assets/57976479/08a803fe-3d56-4cc7-9872-dfc6c6b0ba90)

- [**Issue 2**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784557201)
Problem:
  - Two errors occurred.
  - `gi.actions/group/join` is called twice. It should be called only once.
![image](https://github.com/okTurtles/group-income/assets/57976479/b1dcb2fb-ebc7-449c-bdf5-f9a099ad0cb8)
![image](https://github.com/okTurtles/group-income/assets/57976479/8b10d1f0-4821-4d71-ac30-906ca38b6b6c)
![image](https://github.com/okTurtles/group-income/assets/57976479/1c5ccaea-a7a5-47a5-9108-7247b8530acd)

- [**Issue 3**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784603749)
Problem:
  - Error in [Tab2] after being kicked from the group. (Not before to try join)
  - Three `gi.actions/group/join` calls. Only two calls are needed.
![image](https://github.com/okTurtles/group-income/assets/57976479/35eb0218-3acf-4f4b-9ada-895fd7af0f74)
![image](https://github.com/okTurtles/group-income/assets/57976479/398762c3-f628-4aaf-a4fe-b17171d64b07)

- [**Issue 4**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784699012)
Problem: 
  - 4 errors in console in Tab4
  - `gi.actions/group/join` is called twice. It should be called only once.
![image](https://github.com/okTurtles/group-income/assets/57976479/115dd15b-5d08-4608-b0dc-8f1aea62d5f4)
![image](https://github.com/okTurtles/group-income/assets/57976479/cd296517-fbdb-432f-891b-dbee9ba4f6f3)
- [**Issue 5**](https://github.com/okTurtles/group-income/pull/1780#issuecomment-1784711834)
Seems fine.

(user1 = alexjin, user2 = greg, user3 = andrea)


IMHO, I don't approve the many updates in `model/contracts` and `controllers/actions`. Especially, I can't approve the way to leave chatroom when leaving group.

corrideat committed:
Bugfixes

corrideat committed:
Clear deferredRemove flag

corrideat committed:
Check if we have the decryption key first

corrideat committed:
Updated error message

corrideat committed:
Remove debug logs and fix multiple join calls

corrideat committed:
Remove unnecessary sync of identity contract

corrideat committed:
Bugfixes

corrideat committed:
Fix UI inconsistency issues (wrong message when joining / leaving; switching to newly created chatroom)

corrideat committed:
Encrypt OP_KEY_SHARE envelope

corrideat committed:
Add missing sync when added to chatroom

corrideat committed:
Remove extraneous logging

taoeffect commented on shared/domains/chelonia/internals.js:
```diff
@@ -926,8 +969,8 @@ export default (sbp('sbp/selectors/register', {
       // of waiting, we need to remove it ourselves
       const keyId = findKeyIdByName(contractState, keyName)
       if (!keyId) {
+        console.log('@@@@ pendingWatch --- pushed to keysToDelete', { contractID, externalContractID, keyName, state: cloneDeep(contractState) })
```
Is this logging still needed?

taoeffect reviewed (approved):
This is looking really solid now, thanks @Silver-IT for finding those issues and @corrideat for figuring out the fixes! I think it's ready to merge into `e2e-protocol` but with also double-check with @Silver-IT to see if he can confirm that the issues he found have been fixed.

corrideat committed:
Add destructor to leave chatrooms

corrideat committed:
Leave chatrooms on removeOurselves

corrideat committed:
Remove JOINING_CHATROOM
